hoomd/data/typeparam.py

```diff
@@ -4,12 +4,87 @@
 
 """Implement TypeParameter."""
 
+from collections.abc import MutableMapping
 from hoomd.data.parameterdicts import AttachedTypeParameterDict
 
 
-class TypeParameter:
-    """Store parameters per type."""
-    __slots__ = ('name', 'type_kind', 'param_dict')
+class TypeParameter(MutableMapping):
+    """Implement a type based mutable mapping.
+
+    *Implements the `collections.abc.MutableMapping` interface (``__delitem__``
+    is disallowed).*
+
+    `TypeParameter` instances extend the base Python mapping interface with with
+    smart defaults, value/key validation/processing, and advanced indexing.
```

Comment:
Explain that the class store parameters indexed by type name or pairs of type names.